### PR TITLE
Adapt Dubbo to 3.1.2

### PR DESCRIPTION
**Pull request by AlbumenJ**

### CODE REVIEW

**Review by jasondeng1997 — APPROVED** [review]

File: spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/bootstrap/DubboBootstrapWrapper.java [hunk 1]
```diff
@@ -36,7 +36,7 @@ public static DubboBootstrapWrapper getInstance() {
 
 	public boolean isReady() {
 		return DubboBootstrap.getInstance().isStarted()
-				&& DubboBootstrap.getInstance().isReady()
+				&& DubboBootstrap.getInstance().isStarted()
```
Comment (steverao) [inline spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/bootstrap/DubboBootstrapWrapper.java]: It's duplicate with above line 38.

File: spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/autoconfigure/DubboMetadataBeanConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * Spring Boot Auto-Configuration class for Dubbo Metadata Bean.
+ *
+ * @author <a href="mailto:mercyblitz@gmail.com">Mercy</a>
+ */
+@Configuration(proxyBeanMethods = false)
+public class DubboMetadataBeanConfiguration {
```
Comment (steverao) [inline spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/autoconfigure/DubboMetadataBeanConfiguration.java:41]: Why you need to split them out from DubboMetadataAutoConfiguration?

File: pom.xml [hunk 3]
```diff
@@ -157,7 +157,7 @@
                 </exclusions>
             </dependency>
 
-            <dependency>
+             <dependency>
```
Comment (steverao) [inline pom.xml]: Care about the space!

File: spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/actuate/endpoint/DubboDiscoveryEndpoint.java [hunk 4]
```diff
@@ -16,24 +16,26 @@
...
+import org.apache.dubbo.registry.support.RegistryManager;
 import org.apache.dubbo.rpc.Invoker;
+import org.apache.dubbo.rpc.model.ModuleModel;
 
+import com.alibaba.cloud.dubbo.registry.DubboCloudRegistry;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.actuate.endpoint.annotation.Endpoint;
 import org.springframework.boot.actuate.endpoint.annotation.ReadOperation;
 
+import java.util.ArrayList;
```
Comment (steverao) [inline spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/actuate/endpoint/DubboDiscoveryEndpoint.java]: The order of importing package isn't compliant with the SCA community rule.

File: spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/autoconfigure/DubboMetadataBeanConfiguration.java (hunk 2 above)
Comment (AlbumenJ) [reply spring-cloud-alibaba-starters/spring-cloud-starter-dubbo/src/main/java/com/alibaba/cloud/dubbo/autoconfigure/DubboMetadataBeanConfiguration.java:41]: To resolve circular reference

**Pull request merged**